client/db.go

```diff
@@ -0,0 +1,529 @@
+// TODO(pmattis): ConditionalPut, DeleteRange
+
+package client
+
+import (
+	"encoding"
+	"fmt"
+	"log"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/testutils"
+	gogoproto "github.com/gogo/protobuf/proto"
+)
+
+type byteSliceStringer []byte
+
+func (v byteSliceStringer) String() string {
+	if v == nil {
+		return "nil"
+	}
+	return string(v)
+}
+
+type int64Stringer int64
+
+func (v int64Stringer) String() string {
+	return strconv.FormatInt(int64(v), 10)
+}
+
+// KeyValue ...
+type KeyValue struct {
+	key       []byte
+	value     interface{}
+	timestamp int64
+}
+
+func (kv *KeyValue) String() string {
+	return string(kv.key) + ":" + kv.ValueStringer().String()
+}
+
+// Key ...
+func (kv *KeyValue) Key(k interface{}) error {
+	switch t := k.(type) {
+	case encoding.BinaryUnmarshaler:
+		return t.UnmarshalBinary(kv.key)
+	case *[]byte:
+		*t = kv.key
+	case *string:
+		*t = string(kv.key)
+	}
+	return fmt.Errorf("unable to unmarshal value: %T", k)
+}
+
+// KeyBytes ...
+func (kv *KeyValue) KeyBytes() []byte {
+	return kv.key
+}
+
+// Exists ...
+func (kv *KeyValue) Exists() bool {
+	return kv.value != nil
+}
+
+// Value ...
+func (kv *KeyValue) Value(v interface{}) error {
+	switch val := kv.value.(type) {
+	case []byte:
+		switch t := v.(type) {
+		case encoding.BinaryUnmarshaler:
+			return t.UnmarshalBinary(val)
+		case gogoproto.Message:
+			return gogoproto.Unmarshal(val, t)
+		case *[]byte:
+			*t = val
+		case *string:
+			*t = string(val)
+		}
+	case *int64:
+		switch t := v.(type) {
+		case *int64:
+			*t = *val
+		case *uint64:
+			*t = uint64(*val)
+		case *int32:
+			*t = int32(*val)
+		case *uint32:
+			*t = uint32(*val)
+		case *int:
+			*t = int(*val)
+		case *uint:
+			*t = uint(*val)
+		}
+	}
+	return fmt.Errorf("unable to unmarshal value: %T -> %T", kv.value, v)
+}
+
+func (kv *KeyValue) setValue(v *proto.Value) {
+	if v == nil {
+		return
+	}
+	if v.Bytes != nil {
+		kv.value = v.Bytes
+	} else if v.Integer != nil {
+		kv.value = v.Integer
+	}
+	if ts := v.Timestamp; ts != nil {
+		kv.timestamp = ts.WallTime
+	}
+}
+
+// ValueBytes ...
+func (kv *KeyValue) ValueBytes() []byte {
+	return kv.value.([]byte)
+}
+
+// ValueStringer ...
+func (kv *KeyValue) ValueStringer() fmt.Stringer {
```

Comment:
This seems awkward to me. The name `Stringer`, while standard, is not seen very often in practice, so it feels unfamiliar. It's better to say `string(kv.ValueBytes())` than `kv.ValueStringer.String()`. This is useful when you don't know the underlying type, but when would you need to be type-agnostic but also care about the efficiency of Stringer vs string?